File: leetcode-04/solutions/python3/0870.py
class Solution:
    def advantageCount(self, A: List[int], B: List[int]) -> List[int]:
        A.sort()
        dict = collections.defaultdict(list)

        for b in sorted(B)[::-1]:
            if b < A[-1]:
                dict[b].append(A.pop())

        return [(dict[b] or A).pop() for b in B]

File: leetcode-04/solutions/cpp/0870.cpp
class Solution {
 public:
  vector<int> advantageCount(vector<int>& A, vector<int>& B) {
    multiset<int> set(A.begin(), A.end());

    for (int i = 0; i < B.size(); i++) {
      auto p = *set.rbegin() <= B[i] ? set.begin() : set.upper_bound(B[i]);
      A[i] = *p;
      set.erase(p);
    }

    return A;
  }
};